### ISSUE TITLE
Flaky E2E test TestDeleteServices

### ISSUE THREAD

**Issue by sebgl** [elastic/cloud-on-k8s#2602]
This test failed twice recently:

```
=== RUN   TestDeleteServices/ES_services_should_have_endpoints#01
Retries (5m0s timeout): ....................................................................................................
{"log.level":"error","@timestamp":"2020-02-24T01:06:57.832Z","message":"stopping early","service.version":"0.0.0-00000000","service.type":"eck","ecs.version":"1.4.0","error":"test failure","error.stack_trace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:42\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.runFailureScenario\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/run_failure.go:56\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.RunRecoverableFailureScenario\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/run_failure.go:19\ngithub.com/elastic/cloud-on-k8s/test/e2e/es.TestDeleteServices\n\t/go/src/github.com/elastic/cloud-on-k8s/test/e2e/es/failure_test.go:66\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
    --- FAIL: TestDeleteServices/ES_services_should_have_endpoints#01 (300.00s)
        utils.go:83: 
            	Error Trace:	utils.go:83
            	Error:      	Received unexpected error:
            	            	endpoints "test-failure-delete-services-5zrb-es-http" not found
            	Test:       	TestDeleteServices/ES_services_should_have_endpoints#01
```

Looking at the diagnostic bundle for that E2E test run, I can see that a service `test-failure-delete-services-5zrb-es-http` correctly exists.
According to the operator logs it does not have any endpoint available (which is also reported by the E2E test itself):

```
{"log.level":"error","@timestamp":"2020-02-24T01:04:54.831Z","log.logger":"controller-runtime.controller","message":"Reconciler error","service.version":"1.1.0-a6060c4a","service.type":"eck","ecs.version":"1.4.0","controller":"elasticsearch-controller","request":"e2e-qvv8l-mercury/test-failure-delete-services-5zrb","error":"Endpoints \"test-failure-delete-services-5zrb-es-http\" not found","errorCauses":[{"error":"Endpoints \"test-failure-delete-services-5zrb-es-http\" not found"}],"error.stack_trace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191028221656-72ed19daf4bb/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191028221656-72ed19daf4bb/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191028221656-72ed19daf4bb/pkg/util/wait/wait.go:88"}
```

The Elasticsearch Pod is ready so its IP should be reported in the endpoints. But in this case it looks like the Endpoints resource does not even exist.

**Comment by sebgl** [comment]
I'm wondering if this could be a [bug in Kubernetes](https://github.com/kubernetes/kubernetes/issues/47711).

**Comment by sebgl** [comment]
Happened again: https://devops-ci.elastic.co/job/cloud-on-k8s-e2e-tests-stack-versions/42/testReport/junit/github/com_elastic_cloud-on-k8s_test_e2e_es/Run_tests_for_different_ELK_stack_versions_in_GKE___7_3_2___TestDeleteServices_ES_services_should_have_endpoints_01/

I still think this is a k8s bug.

**Comment by sebgl** [comment]
I've been running this test in a loop locally (agains my GKE cluster): I have not been able to reproduce.

Suspecting a k8s bug I also created a stupid Go script to create/delete a service in a loop and see if I eventually get in a situation where endpoints are not created. No success so far.

```go
// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
// or more contributor license agreements. Licensed under the Elastic License;
// you may not use this file except in compliance with the Elastic License.

package main

import (
	"fmt"
	"time"

	corev1 "k8s.io/api/core/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/apimachinery/pkg/util/intstr"
	"k8s.io/client-go/kubernetes/scheme"
	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp" // allow gcp authentication
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/client/config"

	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
)

// createClient creates a Kubernetes client targeting the current default K8s cluster.
func createClient() (k8s.Client, error) {
	cfg, err := config.GetConfig()
	if err != nil {
		return nil, err
	}
	c, err := client.New(cfg, client.Options{Scheme: scheme.Scheme})
	if err != nil {
		return nil, err
	}
	return k8s.WrapClient(c), nil
}

func main() {
	client, err := createClient()
	if err != nil {
		panic(err)
	}
	for {
		svc := corev1.Service{
			ObjectMeta: metav1.ObjectMeta{
				Namespace: "default",
				Name:      "mysvc",
			},
			Spec: corev1.ServiceSpec{
				Ports: []corev1.ServicePort{
					{
						Name:       "https",
						Port:       9200,
						Protocol:   "TCP",
						TargetPort: intstr.FromInt(9200),
					},
				},
				Selector: map[string]string{
					"common.k8s.elastic.co/type":                "elasticsearch",
					"elasticsearch.k8s.elastic.co/cluster-name": "elasticsearch-sample",
				},
				Type: "ClusterIP",
			},
		}
		fmt.Println("creating svc", svc.Name)
		if err := client.Create(&svc); err != nil {
			panic(err)
		}
		fmt.Println("waiting for endpoints", svc.Name)
		start := time.Now()
		for {
			var endpoints corev1.Endpoints
			err := client.Get(types.NamespacedName{Namespace: "default", Name: "mysvc"}, &endpoints)
			if err != nil {
				if apierrors.IsNotFound(err) {
					fmt.Println(".")
					time.Sleep(5 * time.Second)
					continue
				}
				panic(err)
			}
			fmt.Println("got endpoint after", time.Since(start))
			break
		}
		fmt.Println("deleting svc", svc.Name)
		if err := client.Delete(&svc); err != nil {
			panic(err)
		}
		fmt.Println("waiting for endpoint deletion")
		for {
			var endpoints corev1.Endpoints
			err := client.Get(types.NamespacedName{Namespace: "default", Name: "mysvc"}, &endpoints)
			if err != nil && apierrors.IsNotFound(err) {
				break
			} else if err != nil {
				panic(err)
			}
			fmt.Println(".")
			time.Sleep(2 * time.Second)
		}
	}
	//time.Sleep(10 * time.Second)
}

```

**Comment by sebgl** [comment]
Reproduced on the temp E2E test job (after several hundreds of successful tests): https://devops-ci.elastic.co/job/cloud-on-k8s-e2e-tests-testdeleteservices/667/console

10 hours after the test failed, there are still no endpoints for the `test-failure-delete-services-smr5-es-http` service:

```
⟩ k get svc -n e2e-acsbk-mercury
NAME                                              TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
test-failure-delete-services-smr5-es-http         ClusterIP   10.1.142.223   <none>        9200/TCP   10h
test-failure-delete-services-smr5-es-masterdata   ClusterIP   None           <none>        <none>     10h
test-failure-delete-services-smr5-es-transport    ClusterIP   None           <none>        9300/TCP   10h


⟩ k get endpoints -n e2e-acsbk-mercury
NAME                                              ENDPOINTS         AGE
test-failure-delete-services-smr5-es-masterdata   10.1.145.9        10h
test-failure-delete-services-smr5-es-transport    10.1.145.9:9300   10h
```

**Comment by thbkrkr** [comment]
22h, 7min and 35s after the creation of the service, the endpoint was finally created.

```yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: "2020-05-02T20:47:53Z"
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: test-failure-delete-services-smr5
  name: test-failure-delete-services-smr5-es-http
  namespace: e2e-acsbk-mercury
  resourceVersion: "345500"
  selfLink: /api/v1/namespaces/e2e-acsbk-mercury/endpoints/test-failure-delete-services-smr5-es-http
  uid: 31f1b3dd-8cb6-11ea-a091-42010a9a005e
subsets:
- addresses:
  - ip: 10.1.145.9
    nodeName: gke-eck-debug-endpoints--default-pool-fc152541-7p15
    targetRef:
      kind: Pod
      name: test-failure-delete-services-smr5-es-masterdata-0
      namespace: e2e-acsbk-mercury
      resourceVersion: "3985"
      uid: 9c55b0c0-8bfc-11ea-a091-42010a9a005e
  ports:
  - name: https
    port: 9200
    protocol: TCP
```

**Comment by sebgl** [comment]
Opened a GCP support case so they can hopefully troubleshoot this from the controller-manager logs we don't have access to.

**Comment by sebgl** [comment]
https://github.com/elastic/cloud-on-k8s/pull/3087 changes the testing logic to work around this issue and stop causing too much noise for us due to the flakiness.

I still believe there's a race condition somewhere here in K8s and/or GKE, we continue the investigation with the GCP support team in the meantime, but I believe there's not much we can do in ECK to solve that.

**Comment by thbkrkr** [comment]
Note: the bug in k8s (https://github.com/kubernetes/kubernetes/issues/47711) has been fixed thanks to https://github.com/kubernetes/kubernetes/pull/93030 which should be released in [1.19](https://github.com/kubernetes/kubernetes/pull/93030#issuecomment-671624899).
